### ISSUE TITLE
Weights initialization and support for training restart / curriculum learning

### ISSUE THREAD

**Issue by a-paranjape** [a-paranjape/mlfundas#2]
* Allow weights and biases to be initialized 'by hand', either using arrays or from file, before training commences. Include support to 'resume' training. 
* Also include new modules for [curriculum learning](https://www.researchgate.net/publication/221344862_Curriculum_learning).

**Comment by a-paranjape** [comment]
Initializing weights from file can be done as follows.

First, make the following structural changes:
* In `mlmodules.py`, define a new function `gen_filestems(modules,file_stem)` where `modules` is a list of instantiated modules and `file_stem` is what it says. This function should do what is currently being done, e.g., in the 5 lines starting [here](https://github.com/a-paranjape/mlfundas/blob/29bb6772e61ab580d81ae2e06840098f674cfc75/code/mlalgos.py#L567) and return the updated module list.
* In `mlmodules.py`, update `Modulate` to accept an additional argument `file_stem`. This could be a final kwarg with default `None` (to ensure all updates are completed). At [this line](https://github.com/a-paranjape/mlfundas/blob/29bb6772e61ab580d81ae2e06840098f674cfc75/code/mlmodules.py#L615) outside all loops, call `mod = gen_filestems(mod,file_stem)`.
* The above will ensure that a single call to `Modulate` will correctly set the file stems for each module, including layer/norm strings, once and for all at initialization, instead of repeatedly doing so in each save/load call as is being done currently. To correctly enable this, update *all* occurrences of `Modulate` in `mlalgos.py` with kwarg `file_stem=self.file_stem` (in `Sequential`) and `file_stem=self.file_stem+fs_mod` (two each in `BiSequential` and `GAN`), where `fs_mod` is a file stem modifier string (`_a` or `_w` in `BiSequential`, `_gen` or `_disc` in `GAN`).
* The task done by `gen_filestems` should be deleted from all `save` and `load` methods of `Sequential`, `BiSequential` and `GAN`.

Next, introduce an option for loading weights from file:
* In `mlmodules.py`, update the `Module` class to include a new attribute `self.resume = False` along with others like `self.layer` for compatibility with weighted modules.
* In `mlmodules.py`, update `Modulate` to accept an additional kwarg `resume=False`. Update instantiations of *all* modules with `resume=resume`. 
* In `mlalgos.py`, update `Sequential`, `BiSequential` and `GAN` to
    *  accept a new parameter `'resume'` that sets attribute `self.resume` with default `False`.
    * call `Modulate` with additional kwarg `resume=self.resume`.
* In `mlmodules.py`, update `Linear` and `BatchNorm` to conditionally set `self.W` and `self.W0` with random numbers (as is being done now) when `self.resume==False`, else call `self.load()`. **Note: this means Adam momenta for mean and variance will be reset to zeros with `resume=True`, which is perhaps sensible.**

**Comment by a-paranjape** [comment]
Above won't exactly work since loading existing weights requires file stems to be already set. Instead, port initialization of weights to a new function `initialize_modules` to be called by `Modulate` *after* instantiating all modules along with file stems, i.e. after the call to `gen_filestems()`.

**Comment by a-paranjape** [comment]
Changes made and committed.

**Comment by a-paranjape** [comment]
Basic curriculum learning can be implemented as follows.

Define a new class `CurriculumNetwork`, to be initialized with a choice of keyword `family` (e.g., one of `['seq','biseq','gan']`, ...) and setup parameter dictionary `params` (as appropriate for each family). 

An instance of `CurriculumNetwork` will work as follows:
1. Initializes an instance of `family` as `self.net` with the supplied key `params`.
2. Exposes two methods, `self.train` and `self.predict`:
    * `self.train` has identical call signature as `self.net.train`, except that the `params_train` dict has an additional key 'curriculum' described below:
        * The additional key `curriculum` can be either `None` or a list of integers (slices?) in the range `[1,X.shape[1]]` 
        * When `curriculum` is not `None`, `self.train` splits `X` and `Y` according to the given slices and sequentially reinitializes `self.net` (with `resume=True` starting from the second call), followed by a call to `self.net.train` on each slice for the desired number of epochs. 
        * A final revision phase can consist of a few epochs trained on the full data set.
        * If `curriculum` is `None`, `self.train` simply implements `self.net.train` on the entire sample.
    * `self.predict` is a simple wrapper around `self.net.predict`. This is primarily for use within a script or notebook immediately after training. Once trained, the network can be accessed by loading it like any other instance of the same family.
3. all other methods `XYZ` of `self.net` are also of course available as `self.net.XYZ`.

The idea is to pass `X,Y` in the  `params` dict that are already ordered according to the desired curriculum, with break points indicated in `self.curriculum`.

**Comment by a-paranjape** [comment]
When defined as above, it should also be possible to treat `CurriculumNetwork` as a new family that can be, e.g., optimized by `HyperOpt` and stored as a `NetworkEnsembleObject`.

**Comment by a-paranjape** [comment]
Changes made and committed through merge of branch 'curriculum'.
